src/pure/RealTimeQueue.mo

```diff
@@ -0,0 +1,1026 @@
+    }
+  };
+
+  /// Inspect the optional element on the front end of a queue.
+  /// Returns `null` if `queue` is empty. Otherwise, the front element of `queue`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.pushFront(Queue.pushFront(Queue.empty<Nat>(), 2), 1);
+  ///   assert Queue.peekFront(queue) == ?1;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func peekFront<T>(queue : Queue<T>) : ?T = switch queue {
+    case (#empty) null;
+    case (#one(x)) ?x;
+    case (#two(x, _)) ?x;
+    case (#three(x, _, _)) ?x;
+    case (#idles((l, _), _)) Stacks.first(l);
+    case (#rebal((dir, big, small))) switch dir {
+      case (#left) ?SmallState.peek(small);
+      case (#right) ?BigState.peek(big)
+    }
+  };
+
+  /// Inspect the optional element on the back end of a queue.
+  /// Returns `null` if `queue` is empty. Otherwise, the back element of `queue`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.pushFront(Queue.pushFront(Queue.empty<Nat>(), 2), 1);
+  ///   assert Queue.peekBack(queue) == ?2;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)`.
+  ///
+  /// Space: `O(1)`.
+  public func peekBack<T>(queue : Queue<T>) : ?T = switch queue {
+    case (#empty) null;
+    case (#one(x)) ?x;
+    case (#two(_, y)) ?y;
+    case (#three(_, _, z)) ?z;
+    case (#idles(_, (r, _))) Stacks.first(r);
+    case (#rebal((dir, big, small))) switch dir {
+      case (#left) ?BigState.peek(big);
+      case (#right) ?SmallState.peek(small)
+    }
+  };
+
+  /// Insert a new element on the front end of a queue.
+  /// Returns the new queue with `element` in the front followed by the elements of `queue`.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// persistent actor {
+  ///   let queue = Queue.pushFront(Queue.pushFront(Queue.empty<Nat>(), 2), 1);
+  ///   assert Queue.peekFront(queue) == ?1;
+  ///   assert Queue.peekBack(queue) == ?2;
+  ///   assert Queue.size(queue) == 2;
+  /// }
+  /// ```
+  ///
+  /// Runtime: `O(1)` worst-case!
+  ///
+  /// Space: `O(1)` worst-case!
+  public func pushFront<T>(queue : Queue<T>, element : T) : Queue<T> = switch queue {
+    case (#empty) #one(element);
+    case (#one(y)) #two(element, y);
+    case (#two(y, z)) #three(element, y, z);
+    case (#three(a, b, c)) {
+      let i1 = ((?(element, ?(a, null)), null), 2);
+      let i2 = ((?(c, ?(b, null)), null), 2);
+      #idles(i1, i2)
+    };
+    case (#idles(l0, (r, nR))) {
+      let (l, nL) = Idle.push(l0, element); // enque the element to the left end
+      // check if the size invariant still holds
+      if (3 * nR >= nL) {
+        debug assert 3 * nL >= nR;
+        #idles((l, nL), (r, nR))
```

Comment:
```suggestion
        #idles((l, nL), (r, nR))
```
You could avoid an allocation here of (r, nR). Shame the compiler doesn't do it.

Reply:
(see #307)